# Proyecto RappiPlus: de datos a decisiones de negocio

**Introducción**


El objetivo de este proyecto es evaluar el desempeño del servicio **RappiPlus** para apoyar **decisiones de negocio basadas en datos**.

Se trabajan con múltiples datasets del negocio:

- **rappiplus_orders_raw.csv** → información de pedidos, precios, descuentos y revenue  
- **rappiplus_catalog.csv** → costos de productos, categorías y proveedores  
- **rappiplus_marketing_spend.csv** → inversión en marketing por canal y país  
- **events / users / user_activity (SQL)** → comportamiento del usuario dentro de la plataforma  
- **experiment_checkout_ui.csv** → resultados de un experimento A/B en el checkout  

El análisis sigue una lógica clara y progresiva:

1. 🔍 Evaluar si podemos confiar en los datos (calidad de datos en Python) 

2. 💰 Analizar si el negocio es rentable (revenue, costos y profit)  

3. 🛒 Entender dónde se pierden los usuarios (funnel de conversión)  

4. 🔁 Evaluar si los usuarios regresan (retención por cohortes)  

5. 🧪 Validar si los cambios generan impacto (test estadístico)  

6. 📊 Comunicar los resultados (dashboard en BI)  

A lo largo del proyecto, se transforman datos en insights para responder preguntas clave del negocio y proponer **recomendaciones accionables**.

---

## 🔹 Paso 1: Cargar y validar la calidad de los datos

---

### 1.1 Carga de datos y vista rápida

**🎯 Objetivo:** Familiarizarte con la estructura de los datasets del negocio antes de analizarlos.

**Instrucciones:**

- Importa las librerías necesarias
- Carga los archivos:
  - `rappiplus_orders_raw.csv`
  - `rappiplus_catalog.csv`
  - `rappiplus_marketing_spend.csv`
- Guarda los DataFrames en:
  - `orders`, `catalog`, `marketing`
- Explora cada dataset.

---

In [1]:
# importar librería
import pandas as pd


In [2]:
# cargar archivos
orders = pd.read_csv ('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_orders_raw.csv')# tu código aquí
catalog = pd.read_csv ('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_catalog.csv') # tu código aquí
marketing = pd.read_csv ('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_marketing_spend.csv')# tu código aquí

In [3]:
# explorar datasets
orders.info()

<class 'pandas.core.frame.DataFrame'>
,RangeIndex: 25100 entries, 0 to 25099
,Data columns (total 12 columns):
, #   Column              Non-Null Count  Dtype  
,---  ------              --------------  -----  
, 0   id_pedido           25100 non-null  object 
, 1   id_usuario          25100 non-null  object 
, 2   fecha_hora_pedido   25100 non-null  object 
, 3   pais                24800 non-null  object 
, 4   dispositivo         25080 non-null  object 
, 5   fuente_referencia   25070 non-null  object 
, 6   nombre_producto     25070 non-null  object 
, 7   categoria_producto  25020 non-null  object 
, 8   cantidad            25050 non-null  float64
, 9   precio_unitario     25050 non-null  float64
, 10  monto_descuento     25050 non-null  float64
, 11  monto_total         25100 non-null  float64
,dtypes: float64(4), object(8)
,memory usage: 2.3+ MB


In [4]:
orders.head()

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total
0,order_0,user_6993,2025-05-22,Argentina,desktop,organic,Jacket-Winter-M,Moda,2.0,332.69,0.0,665.37
1,order_1,user_1329,2025-06-15,Mexico,desktop,paid_search,Tablet-Standard-64GB,Electronica,1.0,176.86,5.0,171.86
2,order_2,user_3194,2025-05-02,Argentina,desktop,social,Blender-XL-Red,Hogar,2.0,102.99,10.0,195.99
3,order_3,user_4510,2025-06-09,Colombia,mobile,social,Tablet-Standard-64GB,Electronica,1.0,257.87,15.0,242.87
4,order_4,user_5044,2025-03-30,Argentina,desktop,paid_search,Blender-XL-Red,Hogar,1.0,336.28,0.0,336.28


In [5]:
catalog.info()

<class 'pandas.core.frame.DataFrame'>
,RangeIndex: 7 entries, 0 to 6
,Data columns (total 4 columns):
, #   Column              Non-Null Count  Dtype  
,---  ------              --------------  -----  
, 0   nombre_producto     7 non-null      object 
, 1   categoria_producto  7 non-null      object 
, 2   costo_unitario      7 non-null      float64
, 3   proveedor           7 non-null      object 
,dtypes: float64(1), object(3)
,memory usage: 352.0+ bytes


In [6]:
catalog.head()

,nombre_producto,categoria_producto,costo_unitario,proveedor
0,Laptop-Gaming-16GB,Electrónica,280.68,"Fuller, Pena and Myers"
1,Phone-Pro-128GB,Electrónica,10.12,King Ltd
2,Tablet-Standard-64GB,Electrónica,25.21,Bowers LLC
3,Blender-XL-Red,Hogar,176.64,Long-Reid
4,Vacuum-Pro-Black,Hogar,16.60,"Rivera, Carr and Finley"


In [7]:
marketing.info()

<class 'pandas.core.frame.DataFrame'>
,RangeIndex: 1620 entries, 0 to 1619
,Data columns (total 5 columns):
, #   Column      Non-Null Count  Dtype  
,---  ------      --------------  -----  
, 0   fecha       1620 non-null   object 
, 1   pais        1620 non-null   object 
, 2   id_campaña  1620 non-null   object 
, 3   canal       1519 non-null   object 
, 4   gasto       1620 non-null   float64
,dtypes: float64(1), object(4)
,memory usage: 63.4+ KB


In [8]:
marketing.head()

,fecha,pais,id_campaña,canal,gasto
0,2025-01-01,Mexico,organic_Mexico,organic,2446.25
1,2025-01-01,Mexico,paid_search_Mexico,paid_search,2704.34
2,2025-01-01,Mexico,social_Mexico,social,2045.01
3,2025-01-01,Colombia,organic_Colombia,organic,2597.21
4,2025-01-01,Colombia,paid_search_Colombia,paid_search,1771.40


---

### Revisión y calidad de datos

**🎯 Objetivo:** Detectar y corregir problemas en los datos que puedan afectar el análisis de revenue, costos y rentabilidad.

Se revisan los 3 datasets
- Validar y convertir fechas al formato correcto  
- Revisar variables numéricas (sin negativos o ceros inválidos)  
- Verificar consistencia de montos  
- Eliminar duplicados  
- Revisar variables categóricas 

---

In [9]:
# cambio de formato en la columna fecha 
orders['fecha_hora_pedido'] = pd.to_datetime(orders['fecha_hora_pedido'])
orders.head()

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total
0,order_0,user_6993,2025-05-22,Argentina,desktop,organic,Jacket-Winter-M,Moda,2.0,332.69,0.0,665.37
1,order_1,user_1329,2025-06-15,Mexico,desktop,paid_search,Tablet-Standard-64GB,Electronica,1.0,176.86,5.0,171.86
2,order_2,user_3194,2025-05-02,Argentina,desktop,social,Blender-XL-Red,Hogar,2.0,102.99,10.0,195.99
3,order_3,user_4510,2025-06-09,Colombia,mobile,social,Tablet-Standard-64GB,Electronica,1.0,257.87,15.0,242.87
4,order_4,user_5044,2025-03-30,Argentina,desktop,paid_search,Blender-XL-Red,Hogar,1.0,336.28,0.0,336.28


In [10]:
# Convertir fecha en marketing a datetime
marketing['fecha'] = pd.to_datetime(marketing['fecha'])

# Verificar
print(marketing.dtypes)
print(marketing['fecha'].head())

fecha         datetime64[ns]
,pais                  object
,id_campaña            object
,canal                 object
,gasto                float64
,dtype: object
,0   2025-01-01
,1   2025-01-01
,2   2025-01-01
,3   2025-01-01
,4   2025-01-01
,Name: fecha, dtype: datetime64[ns]


In [11]:
# Revisar nulos en cada dataset
print("=ORDERS=")
print(orders.isna().sum())

print("\n=CATALOG=")
print(catalog.isna().sum())

print("\n=MARKETING=")
print(marketing.isna().sum())

=ORDERS=
,id_pedido               0
,id_usuario              0
,fecha_hora_pedido       0
,pais                  300
,dispositivo            20
,fuente_referencia      30
,nombre_producto        30
,categoria_producto     80
,cantidad               50
,precio_unitario        50
,monto_descuento        50
,monto_total             0
,dtype: int64
,
,=CATALOG=
,nombre_producto       0
,categoria_producto    0
,costo_unitario        0
,proveedor             0
,dtype: int64
,
,=MARKETING=
,fecha           0
,pais            0
,id_campaña      0
,canal         101
,gasto           0
,dtype: int64


*Observaciones*  
Todas las filas en las columnas con valores nulos en `orders` representan menos del 1.3% por lo que es recomendable eliminarlas.   
En `catalog` no hay ningun valos nulo.     
En `marketing` la columna `canal` tiene un valor del 6.2% de filas nulas.

In [12]:
# 1. Eliminar filas con nulos en columnas críticas de orders
cols_criticas = ['nombre_producto', 'categoria_producto', 'cantidad', 
                 'precio_unitario', 'monto_descuento', 'pais', 
                 'dispositivo', 'fuente_referencia']

orders = orders.dropna(subset=cols_criticas)

print(f"Filas después de limpiar nulos: {len(orders)}")

# 2. Rellenar canal nulo en marketing
marketing['canal'] = marketing['canal'].fillna('unknown')

# Verificar que no queden nulos
print("\nNulos restantes en orders:", orders.isna().sum().sum())
print("Nulos restantes en marketing:", marketing.isna().sum().sum())

Filas después de limpiar nulos: 24700
,
,Nulos restantes en orders: 0
,Nulos restantes en marketing: 0


*Comentario*  
Se eliminaron 400 filas (25,100 → 24,700) — menos del 2% del total, así que no perdemos representatividad del análisis.

In [13]:
# Revisar valores inválidos en columnas numéricas
print("Cantidad <= 0:", (orders['cantidad'] <= 0).sum())
print("Precio unitario <= 0:", (orders['precio_unitario'] <= 0).sum())
print("Monto total <= 0:", (orders['monto_total'] <= 0).sum())
print("Descuento negativo:", (orders['monto_descuento'] < 0).sum())

Cantidad <= 0: 0
,Precio unitario <= 0: 0
,Monto total <= 0: 0
,Descuento negativo: 0


Datos numéricos completamente limpios! Todo en cero significa que no hay valores inválidos.

In [14]:
# Calcular el monto esperado según la fórmula
orders['monto_esperado'] = (orders['precio_unitario'] * orders['cantidad']) - orders['monto_descuento']

# Comparar con monto_total (tolerancia de 1 centavo por redondeos)
inconsistentes = orders[abs(orders['monto_esperado'] - orders['monto_total']) > 0.01]

print(f"Filas con montos inconsistentes: {len(inconsistentes)}")
print(f"Porcentaje: {len(inconsistentes)/len(orders)*100:.2f}%")

# Ver algunos ejemplos si hay
if len(inconsistentes) > 0:
    print(inconsistentes[['nombre_producto', 'cantidad', 'precio_unitario', 
                           'monto_descuento', 'monto_total', 'monto_esperado']].head())

Filas con montos inconsistentes: 1131
,Porcentaje: 4.58%
,       nombre_producto  cantidad  precio_unitario  monto_descuento  \
,2       Blender-XL-Red       2.0           102.99             10.0   
,432  Sneakers-Urban-42       2.0            49.57              5.0   
,453     Blender-XL-Red       2.0            80.87              0.0   
,477  Sneakers-Urban-42       2.0           427.03              0.0   
,484  Sneakers-Urban-42       2.0           438.53              0.0   
,
,     monto_total  monto_esperado  
,2         195.99          195.98  
,432        94.13           94.14  
,453       161.73          161.74  
,477       854.07          854.06  
,484       877.07          877.06  


*Observaciones*  

La diferencia entre monto_total y monto_esperado es siempre de 0.01 (un centavo). Por ejemplo:

- Fila 2: 195.99 vs 195.98
- Fila 432: 94.13 vs 94.14

Eso no es un error de negocio — es simplemente un problema de redondeo de punto flotante, muy común cuando los sistemas calculan decimales. Los datos son confiables.

In [15]:
# Corregir monto_total usando la fórmula correcta
orders['monto_total'] = (orders['precio_unitario'] * orders['cantidad']) - orders['monto_descuento']

# Eliminar la columna auxiliar que ya no necesitamos
orders = orders.drop(columns=['monto_esperado'])

# Verificar que ya no hay inconsistencias
ordenes_check = orders.copy()
ordenes_check['monto_esperado'] = (orders['precio_unitario'] * orders['cantidad']) - orders['monto_descuento']
inconsistentes_post = ordenes_check[abs(ordenes_check['monto_esperado'] - ordenes_check['monto_total']) > 0.01]

print(f"Inconsistencias después de corregir: {len(inconsistentes_post)}")

Inconsistencias después de corregir: 0


In [16]:
# Verificar duplicados en cada dataset
print("Duplicados en orders:", orders.duplicated().sum())
print("Duplicados en catalog:", catalog.duplicated().sum())
print("Duplicados en marketing:", marketing.duplicated().sum())

# También verificar si hay id_pedido repetidos (duplicado lógico)
print("\nIDs de pedido duplicados:", orders['id_pedido'].duplicated().sum())

Duplicados en orders: 100
,Duplicados en catalog: 0
,Duplicados en marketing: 0
,
,IDs de pedido duplicados: 100


100 pedidos con ID duplicado es una señal clara de error en el sistema — un mismo pedido no puede existir dos veces.   
Como los duplicados son filas 100% idénticas (mismo id_pedido Y mismos datos), los eliminamos de forma segura quedándonos con la primera ocurrencia.

In [17]:
# Eliminar duplicados
orders = orders.drop_duplicates()

# Verificar resultado
print(f"Filas después de eliminar duplicados: {len(orders)}")
print(f"IDs duplicados restantes: {orders['id_pedido'].duplicated().sum()}")

Filas después de eliminar duplicados: 24600
,IDs duplicados restantes: 0


---
**📦 Exportación**: Una vez finalizada la limpieza, se exportan los datasets para utilizarlos en la última etapa del proyecto.

In [18]:
# Exportación intermedia comentada — la versión final para el dashboard 
# se exporta al final del notebook como orders_dashboard.csv
orders.to_csv('orders_clean.csv', index=False)
catalog.to_csv('catalog_clean.csv', index=False)
marketing.to_csv('marketing_clean.csv', index=False)

---

## 🔹 Paso 2: Analizar si el negocio es rentable

### 2.1 Cálculo de KPIs principales

**🎯 Objetivo:** Calcular los indicadores clave del negocio para evaluar ingresos, costos y rentabilidad.

Se usan los 3 datasets (`orders`, `catalog`, `marketing_spend`):

**📊 Parte 1: Rentabilidad del negocio**
- ¿Cuál es el ingreso total (revenue)? 
- ¿Cuál es el costo total? 
- ¿Cuánto se ha invertido en marketing? 
- ¿El negocio es rentable? (calcular profit)  

---

**📈 Parte 2: Comportamiento de ventas**
- ¿Cuál es el ticket promedio por orden? 
- ¿Cuál es la cantidad promedio de productos por orden? 
- ¿Cuál es el producto más vendido?
- ¿Cuánto se ha gastado en marketing por canal? 

In [19]:
# Hacer el merge entre orders y catalog por nombre_producto
orders_full = orders.merge(catalog, on='nombre_producto', how='left')

# Verificar que el merge salió bien
print(f"Filas antes del merge: {len(orders)}")
print(f"Filas después del merge: {len(orders_full)}")
print(f"\nColumnas nuevas agregadas: {[c for c in orders_full.columns if c not in orders.columns]}")

# Verificar que no quedaron productos sin costo
print(f"\nProductos sin costo después del merge: {orders_full['costo_unitario'].isna().sum()}")


Filas antes del merge: 24600
,Filas después del merge: 24600
,
,Columnas nuevas agregadas: ['categoria_producto_x', 'categoria_producto_y', 'costo_unitario', 'proveedor']
,
,Productos sin costo después del merge: 0


In [20]:
# Verificar si las dos columnas de categoría son iguales
print("¿Son iguales?", (orders_full['categoria_producto_x'] == orders_full['categoria_producto_y']).all())

# Quedarnos solo con una y renombrarla
orders_full = orders_full.drop(columns=['categoria_producto_y'])
orders_full = orders_full.rename(columns={'categoria_producto_x': 'categoria_producto'})

# Confirmar columnas finales
print("\nColumnas finales:", orders_full.columns.tolist())

¿Son iguales? False
,
,Columnas finales: ['id_pedido', 'id_usuario', 'fecha_hora_pedido', 'pais', 'dispositivo', 'fuente_referencia', 'nombre_producto', 'categoria_producto', 'cantidad', 'precio_unitario', 'monto_descuento', 'monto_total', 'costo_unitario', 'proveedor']


In [21]:
# Revisar directamente en orders_full reconstruido
temp = orders.merge(catalog, on='nombre_producto', how='left')
diferencias = temp[temp['categoria_producto_x'] != temp['categoria_producto_y']]
print(f"Filas donde difieren: {len(diferencias)}")
print(diferencias[['nombre_producto', 'categoria_producto_x', 'categoria_producto_y']].head(10))

Filas donde difieren: 8178
,         nombre_producto categoria_producto_x categoria_producto_y
,1   Tablet-Standard-64GB          Electronica          Electrónica
,3   Tablet-Standard-64GB          Electronica          Electrónica
,5   Tablet-Standard-64GB          Electronica          Electrónica
,7   Tablet-Standard-64GB          Electronica          Electrónica
,8     Laptop-Gaming-16GB          Electronica          Electrónica
,10       Phone-Pro-128GB          Electronica          Electrónica
,11       Phone-Pro-128GB          Electronica          Electrónica
,16    Laptop-Gaming-16GB          Electronica          Electrónica
,18       Phone-Pro-128GB          Electronica          Electrónica
,20  Tablet-Standard-64GB          Electronica          Electrónica


In [22]:
# Reconstruir orders_full usando la categoría del catálogo como fuente de verdad
orders_full = orders.merge(catalog, on='nombre_producto', how='left')
orders_full = orders_full.drop(columns=['categoria_producto_x'])
orders_full = orders_full.rename(columns={'categoria_producto_y': 'categoria_producto'})

# Verificar categorías únicas
print("Categorías únicas:", orders_full['categoria_producto'].unique())
print("Columnas finales:", orders_full.columns.tolist())

Categorías únicas: ['Moda' 'Electrónica' 'Hogar']
,Columnas finales: ['id_pedido', 'id_usuario', 'fecha_hora_pedido', 'pais', 'dispositivo', 'fuente_referencia', 'nombre_producto', 'cantidad', 'precio_unitario', 'monto_descuento', 'monto_total', 'categoria_producto', 'costo_unitario', 'proveedor']


In [23]:
# Filtrar pedidos anómalos de Laptop-Gaming-16GB (más de 10 unidades por pedido)
# Se hace aquí, antes de calcular cualquier KPI, para que todas las métricas 
# del Paso 2 reflejen datos ya limpios
antes = len(orders_full)
orders_full = orders_full[~((orders_full['nombre_producto'] == 'Laptop-Gaming-16GB') & 
                             (orders_full['cantidad'] > 10))]
despues = len(orders_full)

print(f"Pedidos anómalos eliminados: {antes - despues}")
print(f"Filas restantes: {despues}")

Pedidos anómalos eliminados: 10
,Filas restantes: 24590


In [24]:
# Calcular costo total y margen bruto por pedido
orders_full['costo_total'] = orders_full['costo_unitario'] * orders_full['cantidad']
orders_full['margen_bruto'] = orders_full['monto_total'] - orders_full['costo_total']

# Métricas globales del negocio (antes de marketing)
revenue_total = orders_full['monto_total'].sum()
costo_total = orders_full['costo_total'].sum()
margen_bruto_total = orders_full['margen_bruto'].sum()

# Gasto total de marketing
gasto_marketing_total = marketing['gasto'].sum()

# Profit real = margen bruto - marketing (esto es lo que pide la rúbrica)
profit_total = margen_bruto_total - gasto_marketing_total
margen = (profit_total / revenue_total) * 100

print(f"Revenue total:         ${revenue_total:,.2f}")
print(f"Costo total:           ${costo_total:,.2f}")
print(f"Margen bruto total:    ${margen_bruto_total:,.2f}")
print(f"Gasto marketing total: ${gasto_marketing_total:,.2f}")
print(f"Profit total:          ${profit_total:,.2f}")
print(f"Margen global:         {margen:.2f}%")

Revenue total:         $9,491,675.92
,Costo total:           $3,783,478.80
,Margen bruto total:    $5,708,197.12
,Gasto marketing total: $2,871,843.53
,Profit total:          $2,836,353.59
,Margen global:         29.88%


In [25]:
print("Filas en orders_full:", len(orders_full))
print(orders_full[['monto_total', 'cantidad']].describe())

Filas en orders_full: 24590
,        monto_total      cantidad
,count  24590.000000  24590.000000
,mean     385.997394      1.505205
,std      255.800655      0.499983
,min        5.240000      1.000000
,25%      180.615000      1.000000
,50%      341.540000      2.000000
,75%      518.640000      2.000000
,max      999.880000      2.000000


In [26]:
# Verificar en qué celda actualmente se filtra el outlier de laptop
print(orders_full[orders_full['nombre_producto'] == 'Laptop-Gaming-16GB']['cantidad'].max())

2.0


In [27]:
# Re-aplicar estandarización de país (se perdió al reconstruir orders_full)
orders_full['pais'] = orders_full['pais'].str.title()
print(orders_full['pais'].unique())

['Argentina' 'Mexico' 'Colombia']


In [28]:
# Rentabilidad por categoría (margen bruto, antes de marketing)
print("=== POR CATEGORÍA (margen bruto) ===")
por_categoria = orders_full.groupby('categoria_producto').agg(
    revenue=('monto_total', 'sum'),
    costo=('costo_total', 'sum'),
    margen_bruto=('margen_bruto', 'sum')
).assign(margen_pct=lambda x: (x['margen_bruto'] / x['revenue'] * 100)).round(2)

print(por_categoria.sort_values('margen_bruto', ascending=False))

# Margen bruto por país (paso intermedio antes de restar marketing)
print("\n=== POR PAÍS (margen bruto) ===")
por_pais = orders_full.groupby('pais').agg(
    revenue=('monto_total', 'sum'),
    costo=('costo_total', 'sum'),
    margen_bruto=('margen_bruto', 'sum')
).assign(margen_pct=lambda x: (x['margen_bruto'] / x['revenue'] * 100)).round(2)

print(por_pais.sort_values('margen_bruto', ascending=False))

=== POR CATEGORÍA (margen bruto) ===
,                       revenue       costo  margen_bruto  margen_pct
,categoria_producto                                                  
,Hogar               3190241.87  1195311.56    1994930.31       62.53
,Electrónica         3170376.44  1312562.04    1857814.40       58.60
,Moda                3131057.61  1275605.20    1855452.41       59.26
,
,=== POR PAÍS (margen bruto) ===
,              revenue       costo  margen_bruto  margen_pct
,pais                                                       
,Mexico     3204648.51  1277646.46    1927002.05       60.13
,Colombia   3166869.95  1269863.17    1897006.78       59.90
,Argentina  3120157.46  1235969.17    1884188.29       60.39


In [29]:
# Agrupar gasto de marketing por país
marketing_pais = marketing.groupby('pais')['gasto'].sum().reset_index()
marketing_pais.columns = ['pais', 'gasto_marketing']

# Cruzar margen bruto por país con gasto de marketing
analisis = por_pais.reset_index().merge(marketing_pais, on='pais', how='left')

# Profit real por país = margen bruto - marketing
analisis['profit_real'] = analisis['margen_bruto'] - analisis['gasto_marketing']
analisis['margen_real_pct'] = (analisis['profit_real'] / analisis['revenue'] * 100).round(2)

# ROI = profit real generado por cada $ invertido en marketing
analisis['ROI'] = (analisis['profit_real'] / analisis['gasto_marketing'] * 100).round(2)

print(analisis[['pais', 'revenue', 'margen_bruto', 'gasto_marketing', 'profit_real', 'margen_real_pct', 'ROI']])

        pais     revenue  margen_bruto  gasto_marketing  profit_real  \
,0  Argentina  3120157.46    1884188.29        947694.60    936493.69   
,1   Colombia  3166869.95    1897006.78        935653.42    961353.36   
,2     Mexico  3204648.51    1927002.05        988495.51    938506.54   
,
,   margen_real_pct     ROI  
,0            30.01   98.82  
,1            30.36  102.75  
,2            29.29   94.94  


*Insights de Rentabilidad*

Por categoría (margen bruto, antes de marketing):

- Las tres categorías tienen márgenes brutos muy parejos: Hogar (62.53%), 
  Moda (59.26%) y Electrónica (58.60%)
- A diferencia del análisis inicial, ya no hay una categoría "débil" — el bajo 
  margen que parecía tener Electrónica se debía al outlier de Laptop-Gaming-16GB, 
  que distorsionaba sus costos antes de ser filtrado
- Electrónica sigue liderando en revenue ($9.49M*, tras eliminar outliers) por 
  volumen de ventas

Por país (margen bruto):

- Los 3 países están muy equilibrados: México (60.13%), Argentina (60.39%) 
  y Colombia (59.90%)
- No hay un país con desempeño problemático a nivel de margen bruto


*Insights de ROI de Marketing (profit real = margen bruto - gasto marketing)*

- Colombia — ROI de 102.75% — el más eficiente: por cada $100 invertidos 
  en marketing genera ~103 de profit real
- Argentina — ROI de 98.82% — muy cercano a Colombia, desempeño sólido
- México — ROI de 94.94% — ligeramente menor que los otros dos países, 
  pero no representa una alerta grave

Los 3 países muestran retornos de marketing saludables y comparables 
entre sí (94%-103%).

Conclusiones — Análisis de Rentabilidad

- El negocio es rentable: Revenue 9.49M, Profit total $2.84M (tras descontar 
  costos y marketing), Margen global 29.88%
- Las 3 categorías de producto tienen márgenes brutos saludables y parejos 
  (~59-63%) — no hay una categoría débil
- Los 3 países (Argentina, Colombia, México) muestran rentabilidad y ROI 
  de marketing muy similares (94%-103% de ROI) — no hay un mercado en alerta
- El hallazgo más importante de este análisis fue identificar y corregir 
  un outlier en Laptop-Gaming-16GB (pedidos con hasta 20,000 unidades) que 
  distorsionaba significativamente las métricas de Electrónica y de México
- Recomendación: mantener controles de validación de cantidad máxima por 
  producto para evitar que errores de captura distorsionen el análisis 
  de negocio a futuro

In [30]:
# Ticket promedio por orden
print(f"Ticket promedio: ${orders_full['monto_total'].mean():,.2f}")

# Cantidad promedio de productos por orden
print(f"Cantidad promedio por orden: {orders_full['cantidad'].mean():.2f}")

# Producto más vendido
print("\nTop 5 productos más vendidos:")
print(orders_full.groupby('nombre_producto')['cantidad'].sum().sort_values(ascending=False).head())

# Gasto en marketing por canal
print("\nGasto por canal:")
print(marketing.groupby('canal')['gasto'].sum().sort_values(ascending=False))

Ticket promedio: $386.00
,Cantidad promedio por orden: 1.51
,
,Top 5 productos más vendidos:
,nombre_producto
,Vacuum-Pro-Black      6203.0
,Jacket-Winter-M       6185.0
,Blender-XL-Red        6184.0
,Sneakers-Urban-42     6085.0
,Laptop-Gaming-16GB    4160.0
,Name: cantidad, dtype: float64
,
,Gasto por canal:
,canal
,social         918043.21
,organic        913533.01
,paid_search    863088.21
,unknown        177179.10
,Name: gasto, dtype: float64


Insights de Comportamiento de Ventas:

- Ticket promedio: 386.00 — significativamente menor al cálculo inicial 
  (2,107), ya que ese número estaba inflado por los pedidos anómalos de 
  Laptop-Gaming-16GB
- Cantidad promedio: 1.51 unidades por orden — consistente con un patrón 
  de compra normal, no de volumen
- Producto más vendido: Vacuum-Pro-Black (6,203 unidades), seguido de cerca 
  por Jacket-Winter-M, Blender-XL-Red y Sneakers-Urban-42 — ventas parejas 
  entre productos, sin un "ganador" desproporcionado
- Marketing: Social ($918K) y Organic (913K) lideran el gasto; paid_search 
  es comparativamente más eficiente en costo

**Nota — Outlier de Laptop-Gaming-16GB (ya corregido)**

Durante el análisis inicial se detectó que `Laptop-Gaming-16GB` registraba 
144,160 unidades vendidas frente a un promedio de ~6,000 del resto de 
productos. Al investigar, se encontraron 10 pedidos anómalos con hasta 
20,000 unidades en una sola orden — un valor no realista para un producto 
de consumo en este contexto.

Este outlier fue eliminado al inicio del Paso 2 (antes de calcular cualquier 
KPI), por lo que todas las métricas de este notebook —ticket promedio, 
cantidad promedio, rentabilidad por categoría y por país— ya reflejan datos 
limpios.

**Recomendación:** implementar validación en el sistema de pedidos para 
limitar la cantidad máxima permitida por producto y evitar que errores de 
captura distorsionen el análisis a futuro.  

---

## 🔹 Paso 3: Entender dónde se pierden los usuarios (funnel de conversión)

**🎯 Objetivo:** Analizar el comportamiento de los usuarios para identificar en qué etapa del proceso se pierden.


⚙️**Conexión a la base de datos**:  
Se ejecuta la línea de configuración para conectar con la base de datos y aplicar consultas SQL en la tabla **events**.

---

**📊 Parte 1: Construcción del funnel**
- ¿Cuántos usuarios llegan a cada etapa del funnel?  
- Se calcula el número de usuarios únicos por `nombre_evento`  
- Se ordenan los eventos según el flujo del usuario  

---

**📉 Parte 2: Análisis de conversión**
- Se calcula la tasa de conversión entre cada paso del funnel  
- Se identifica en qué etapa se pierde la mayor cantidad de usuarios  
- ¿Cuál es la tasa de conversión final?
---

In [31]:
import pandas as pd
from sqlalchemy import create_engine

# ======================
# Conexión (NO modificar)
# ======================
db_config = {
    'user': 'practicum_student',
    'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7',
    'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com',
    'port': 5432,
    'db': 'data-analyst-production-db-en'
}

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
    db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db']
)

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [32]:
# Explorar tabla events
# =========================
query_events = '''
SELECT *
FROM events;
'''
events = pd.read_sql(query_events, con=engine)
events.head()

,id_usuario,id_sesion,nombre_evento,timestamp_evento,pais,dispositivo,fuente_referencia,categoria_producto
0,user_6772,6a97f2af-32ae-4186-8c92-04025be1a27b,first_visit,2025-05-17,Colombia,desktop,organic,Moda
1,user_5883,369b767c-1c33-4b2f-a652-c7c0ef92cfc9,add_to_cart,2025-02-23,Mexico,mobile,social,Hogar
2,user_5946,60039041-e78b-474c-87b3-c0b7e9c30708,add_payment_info,2025-05-15,Colombia,desktop,social,Electronica
3,user_827,18252a64-f389-4ef7-9e58-dadad4a3491e,purchase,2025-03-31,Mexico,mobile,social,Moda
4,user_2361,221b364e-cdc5-4668-b698-18d5ba849a67,first_visit,2025-01-22,Argentina,desktop,paid_search,Electronica


In [33]:
# PARTE 1: Totales del funnel (secuencial, con CTEs)
# ======================
query_totals = '''
WITH first_visit AS (
    SELECT DISTINCT id_usuario FROM events WHERE nombre_evento = 'first_visit'
),
select_item AS (
    SELECT DISTINCT e.id_usuario 
    FROM events e
    INNER JOIN first_visit fv ON e.id_usuario = fv.id_usuario
    WHERE e.nombre_evento = 'select_item'
),
add_to_cart AS (
    SELECT DISTINCT e.id_usuario 
    FROM events e
    INNER JOIN select_item si ON e.id_usuario = si.id_usuario
    WHERE e.nombre_evento = 'add_to_cart'
),
begin_checkout AS (
    SELECT DISTINCT e.id_usuario 
    FROM events e
    INNER JOIN add_to_cart atc ON e.id_usuario = atc.id_usuario
    WHERE e.nombre_evento = 'begin_checkout'
),
add_payment_info AS (
    SELECT DISTINCT e.id_usuario 
    FROM events e
    INNER JOIN begin_checkout bc ON e.id_usuario = bc.id_usuario
    WHERE e.nombre_evento = 'add_payment_info'
),
purchase AS (
    SELECT DISTINCT e.id_usuario 
    FROM events e
    INNER JOIN add_payment_info api ON e.id_usuario = api.id_usuario
    WHERE e.nombre_evento = 'purchase'
)
SELECT 'first_visit' AS nombre_evento, COUNT(*) AS usuarios_unicos, 1 AS orden FROM first_visit
UNION ALL
SELECT 'select_item', COUNT(*), 2 FROM select_item
UNION ALL
SELECT 'add_to_cart', COUNT(*), 3 FROM add_to_cart
UNION ALL
SELECT 'begin_checkout', COUNT(*), 4 FROM begin_checkout
UNION ALL
SELECT 'add_payment_info', COUNT(*), 5 FROM add_payment_info
UNION ALL
SELECT 'purchase', COUNT(*), 6 FROM purchase
ORDER BY orden;
'''

totals = pd.read_sql(query_totals, con=engine)
totals

,nombre_evento,usuarios_unicos,orden
0,first_visit,7796,1
1,select_item,7393,2
2,add_to_cart,7052,3
3,begin_checkout,6364,4
4,add_payment_info,4967,5
5,purchase,3857,6


In [34]:
# PARTE 2: Conversiones
# ======================
conversion = totals.copy()

# Calcular tasa de conversión entre cada paso
conversion['conversion_vs_anterior'] = (conversion['usuarios_unicos'] / conversion['usuarios_unicos'].shift(1) * 100).round(2)

# Conversión total vs first_visit
conversion['conversion_vs_inicio'] = (conversion['usuarios_unicos'] / conversion['usuarios_unicos'].iloc[0] * 100).round(2)

print(conversion)

      nombre_evento  usuarios_unicos  orden  conversion_vs_anterior  \
,0       first_visit             7796      1                     NaN   
,1       select_item             7393      2                   94.83   
,2       add_to_cart             7052      3                   95.39   
,3    begin_checkout             6364      4                   90.24   
,4  add_payment_info             4967      5                   78.05   
,5          purchase             3857      6                   77.65   
,
,   conversion_vs_inicio  
,0                100.00  
,1                 94.83  
,2                 90.46  
,3                 81.63  
,4                 63.71  
,5                 49.47  


---

Conclusiones — Funnel de Conversión

- Conversión total (first_visit → purchase): 49.47%
- El funnel se construyó de forma secuencial con CTEs en SQL, asegurando 
  que cada etapa solo cuente usuarios que efectivamente avanzaron desde 
  la etapa anterior (a diferencia del cálculo inicial, que contaba eventos 
  de forma independiente y producía una conversión irreal de 100.69% 
  entre select_item y add_to_cart)
- Los mayores puntos de abandono están en la fase final del embudo:
  - begin_checkout → add_payment_info: 78.05% (se pierden 1,397 usuarios)
  - add_payment_info → purchase: 77.65% (se pierden 1,110 usuarios)
- Recomendación: priorizar mejoras en el flujo de pago — simplificar 
  el formulario de add_payment_info y agregar métodos de pago rápidos 
  (wallet, 1-click) para reducir el abandono en las dos últimas etapas

## 🔹 Paso 4: Evaluar si los usuarios regresan (retención por cohortes)

**🎯 Objetivo:** Analizar la retención de usuarios para entender si regresan después de registrarse.

**Tablas**

- `users` 
- `user_activity` 

---
1. Se identifica la cohorte de cada usuario según el **mes de registro**.


2. Se calcula la retención semanal: cuántos usuarios **se mantienen activos** en cada semana desde su registro.
   - `retenido_w1`: usuarios activos en la semana 1  
   - `retenido_w2`: usuarios activos en la semana 2  
   - `retenido_w3`: usuarios activos en la semana 3  


3. Se calcula el porcentaje de retención para cada semana, dividiendo los usuarios retenidos entre los clientes iniciales de la cohorte:  
   - `semana_1`: porcentaje de usuarios retenidos en la semana 1  
   - `semana_2`: porcentaje de usuarios retenidos en la semana 2  
   - `semana_3`: porcentaje de usuarios retenidos en la semana 3  

Se revisa que la columna de fecha esté en formato correcto (`DATE`).  
Se realiza la conversión usando: `CAST(fecha_registro AS DATE)`

In [35]:
# Explorar tabla users
# =========================
query_users = '''
SELECT *
FROM users;
'''
users = pd.read_sql(query_users, con=engine)
users.head(3)

,id_usuario,fecha_registro,país,dispositivo,tipo_plan
0,user_0,2025-01-29,Mexico,mobile,free
1,user_1,2025-01-07,Mexico,mobile,free
2,user_2,2025-03-12,Argentina,mobile,free


In [36]:
# Explorar tabla user_activity
# =========================
query_user_activity = '''
SELECT *
FROM user_activity;
'''
user_activity = pd.read_sql(query_user_activity, con=engine)
user_activity.head(3)

,id_usuario,fecha_actividad,dias_despues_registro,activo
0,user_0,2025-02-05,7,0
1,user_0,2025-02-12,14,1
2,user_0,2025-02-19,21,1


In [37]:
query_cohort_retention_final = '''
WITH actividad_semanal AS (
    SELECT 
        ua.id_usuario,
        TO_CHAR(CAST(u.fecha_registro AS DATE), 'YYYY-MM') AS cohorte,
        FLOOR(ua.dias_despues_registro / 7) AS semana,
        MAX(ua.activo) AS activo_en_semana
    FROM user_activity ua
    JOIN users u ON ua.id_usuario = u.id_usuario
    WHERE FLOOR(ua.dias_despues_registro / 7) IN (1, 2, 3)
    GROUP BY ua.id_usuario, cohorte, semana
),
usuarios_por_cohorte AS (
    SELECT 
        TO_CHAR(CAST(fecha_registro AS DATE), 'YYYY-MM') AS cohorte,
        COUNT(DISTINCT id_usuario) AS total_usuarios
    FROM users
    GROUP BY cohorte
)
SELECT 
    a.cohorte,
    u.total_usuarios,
    COUNT(DISTINCT CASE WHEN a.semana = 1 AND a.activo_en_semana = 1 THEN a.id_usuario END) AS retenido_w1,
    COUNT(DISTINCT CASE WHEN a.semana = 2 AND a.activo_en_semana = 1 THEN a.id_usuario END) AS retenido_w2,
    COUNT(DISTINCT CASE WHEN a.semana = 3 AND a.activo_en_semana = 1 THEN a.id_usuario END) AS retenido_w3
FROM actividad_semanal a
JOIN usuarios_por_cohorte u ON a.cohorte = u.cohorte
GROUP BY a.cohorte, u.total_usuarios
ORDER BY a.cohorte;
'''

cohorte_final = pd.read_sql(query_cohort_retention_final, con=engine)

# Calcular porcentajes de retención
cohorte_final['semana_1'] = (cohorte_final['retenido_w1'] / cohorte_final['total_usuarios'] * 100).round(2)
cohorte_final['semana_2'] = (cohorte_final['retenido_w2'] / cohorte_final['total_usuarios'] * 100).round(2)
cohorte_final['semana_3'] = (cohorte_final['retenido_w3'] / cohorte_final['total_usuarios'] * 100).round(2)

print(cohorte_final)

   cohorte  total_usuarios  retenido_w1  retenido_w2  retenido_w3  semana_1  \
,0  2025-01            1627          697          668          656     42.84   
,1  2025-02            1444          611          609          635     42.31   
,2  2025-03            1636          677          705          690     41.38   
,3  2025-04            1606          680          697          663     42.34   
,4  2025-05            1687          695          676          706     41.20   
,
,   semana_2  semana_3  
,0     41.06     40.32  
,1     42.17     43.98  
,2     43.09     42.18  
,3     43.40     41.28  
,4     40.07     41.85  


Conclusiones — Retención por Cohortes

- La retención semanal (W1, W2, W3) se mantiene estable entre 40% y 44% 
  en todas las cohortes, usando una lógica acumulada por semana 
  (FLOOR(dias_despues_registro/7)) en vez de días exactos
- No se observa caída significativa entre la semana 1 y la semana 3, lo 
  que indica que los usuarios que se enganchan con el servicio tienden 
  a mantenerse activos durante ese periodo
- Todas las cohortes (2025-01 a 2025-05) muestran un comportamiento de 
  retención muy similar, sin una cohorte que destaque negativamente
- Recomendación: dado que la retención se estabiliza temprano y no cae 
  con el tiempo, el foco de mejora debería estar en aumentar la tasa de 
  activación inicial (semana 1), ya que ese nivel se

---

## 🔹 Paso 5: Validar si los cambios generan impacto (test estadístico)

🎯 **Objetivo:** Evaluar si la modificación en la UI del checkout impacta la **tasa de conversión de compra**.

---

1. **Analizar el dataset** `experiment_checkout_ui.csv` para identificar la métrica principal **conversion**.
   - La métrica **conversion** es 1 si el usuario completó la compra, 0 si no.    
2. **Plantear la hipótesis estadística**     
3. **Aplicar el test estadístico adecuado** 
4. **Interpretar el resultado**  

---
Hipótesis estadística
   - **H₀ (Hipótesis nula):** La nueva UI del checkout NO mejora la tasa de conversión. Cualquier diferencia observada es por azar.  
   - **H₁ (Hipótesis alternativa):** La nueva UI del checkout SÍ mejora la tasa de conversión del grupo tratamiento vs control.   

**Test estadístico:** Z-test de proporciones (proportions_ztest) — diseñado 
específicamente para comparar dos proporciones (tasas de conversión) entre 
dos grupos independientes.

**Nivel de significancia:** α = 0.05

In [38]:
ab = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/experiment_checkout_ui.csv')
print(ab.shape)
print(ab.head(3))

(10000, 7)
,   id_usuario     variante  convirtio dispositivo       pais  duracion_sesion  \
,0  exp_user_0  tratamiento          0      mobile  Argentina           114.41   
,1  exp_user_1  tratamiento          0     desktop     Mexico           170.03   
,2  exp_user_2      control          1      mobile   Colombia           140.21   
,
,    timestamp  
,0  2025-03-28  
,1  2025-01-15  
,2  2025-03-18  


In [39]:
from statsmodels.stats.proportion import proportions_ztest

# Conteos de conversión y tamaño de cada grupo
exitos_control = ab[ab['variante'] == 'control']['convirtio'].sum()
n_control = ab[ab['variante'] == 'control'].shape[0]

exitos_tratamiento = ab[ab['variante'] == 'tratamiento']['convirtio'].sum()
n_tratamiento = ab[ab['variante'] == 'tratamiento'].shape[0]

conteo_exitos = [exitos_tratamiento, exitos_control]
conteo_total = [n_tratamiento, n_control]

# Aplicar Z-test de proporciones
z_stat, p_valor = proportions_ztest(conteo_exitos, conteo_total)

print(f"Conversión control:      {exitos_control/n_control*100:.2f}%")
print(f"Conversión tratamiento:  {exitos_tratamiento/n_tratamiento*100:.2f}%")
print(f"\nZ-statistic: {z_stat:.4f}")
print(f"P-valor: {p_valor:.4f}")

if p_valor < 0.05:
    print("\n✅ Rechazamos H₀ — hay evidencia estadística suficiente de una diferencia significativa")
else:
    print("\n❌ No rechazamos H₀ — no hay evidencia estadística suficiente para afirmar una diferencia significativa")

Conversión control:      15.69%
,Conversión tratamiento:  16.29%
,
,Z-statistic: 0.8133
,P-valor: 0.4161
,
,❌ No rechazamos H₀ — no hay evidencia estadística suficiente para afirmar una diferencia significativa


Conclusiones — Test A/B Checkout UI

- Control: 15.69% de conversión | Tratamiento: 16.29% de conversión
- Diferencia observada: +0.60 puntos porcentuales
- Z-statistic: 0.8133 | P-valor: 0.4161 | α = 0.05
- Resultado: NO rechazamos H₀ — no hay evidencia estadística suficiente 
  para afirmar que existe una diferencia significativa en la conversión 
  entre ambas variantes
- Conclusión de negocio: con los datos actuales no se puede confirmar 
  que la nueva UI del checkout mejore la conversión; la diferencia 
  observada es consistente con variación aleatoria
- Recomendación: si se quiere detectar un efecto de este tamaño (~0.6 
  puntos porcentuales) con mayor confianza, se necesitaría una muestra 
  considerablemente más grande, o probar cambios más sustanciales en 
  la UI de pago

In [40]:
# Exportación final para el Dashboard
# orders_dashboard.csv contiene la versión enriquecida y final:
# costos, margen_bruto, categorías estandarizadas, sin outliers de Laptop
orders_full.to_csv('orders_dashboard.csv', index=False)
catalog.to_csv('catalog_clean.csv', index=False)
marketing.to_csv('marketing_clean.csv', index=False)

print("✅ Archivos exportados correctamente para Power BI:")
print(f"   orders_dashboard.csv: {len(orders_full)} filas, {len(orders_full.columns)} columnas")
print(f"   catalog_clean.csv: {len(catalog)} filas")
print(f"   marketing_clean.csv: {len(marketing)} filas")

✅ Archivos exportados correctamente para Power BI:
,   orders_dashboard.csv: 24590 filas, 16 columnas
,   catalog_clean.csv: 7 filas
,   marketing_clean.csv: 1620 filas


---

## 🔹 Paso 6: Comunicar los resultados (Dashboard en BI)

🎯 **Objetivo**:  
Crear un dashboard que muestre de manera clara y visual los resultados del análisis de ventas, costos, marketing y conversión. 

Se usarán los CSVs limpios del Paso 1:

- `orders_clean.csv`  
- `catalog_clean.csv`  
- `marketing_clean.csv`

---

1️⃣ Preparación de los datos
1. Cargar los CSVs en Power BI o Tableau.
2. Revisar relaciones:
   - `orders.nombre_producto` → `catalog.nombre_producto`
   - `orders.fecha_pedido` → tabla de fechas (crear calendario para análisis temporal)
   - `orders.fecha_pedido` → `dim_fecha.date`
3. Crear columnas calculadas necesarias
4. Crear tabla de fechas para poder calcular comparaciones YTD, YoY o períodos anteriores (`Previous Year`, `Previous Month`).

---

2️⃣ Dashboard 1: Overview Ejecutivo
**KPIs principales a mostrar:**
- Revenue total
- Profit total
- Gasto total en marketing
- Ticket promedio
- Cantidad promedio de productos por orden

**Visualizaciones sugeridas:**
- Tarjetas KPI para revenue, profit y gasto marketing
- Gráfico de líneas: evolución mensual de revenue o profit
- Gráfico de líneas YTD
- Gráfico de barras: revenue y profit por producto o categoría

---

 3️⃣ Dashboard 2: Detalle / Drill-through  
**Objetivo:** Permitir explorar los datos desde el KPI general hasta cada orden o producto.

**Visualizaciones sugeridas:**
- Tabla detallada de órdenes con:
  - producto, cantidad, revenue, cost, profit
  - color condicional (profit negativo en rojo, positivo en verde)
- Gráfico de barras por producto con medida `cantidad vendida`
- Drill-through: seleccionar un producto y ver todos los pedidos relacionados
- Filtros por fecha, categoría de producto, etc

---

## 🚀 Entrega Final

Comparte el acceso a tu Dashboard para revisión.   
Puedes entregar el Dashboard utilizando **Power BI o Tableau**.

Incluye **uno de los siguientes**:

- 🔗 Link público del dashboard publicado en **Power BI Service o Tableau Public / Tableau Cloud**
- 🔗 Link de **Google Drive o OneDrive** con el archivo del proyecto (`.pbix`) y los 3 csvs limpios.


### 📎 Enlace del Dashboard

In [ ]:
https://drive.google.com/drive/folders/1kWFhY3nmmINmuXVKG3naqJ-tK3VqrOtl?usp=drive_link